dev/fattest.simplicity/src/componenttest/security/utils/SSLUtils.java

```diff
@@ -0,0 +1,175 @@
+
+        Calendar calendar = Calendar.getInstance();
+        calendar.setTime(startDate);
+        // 1 Yr validity
+        calendar.add(Calendar.YEAR, 1);
+
+        Date endDate = calendar.getTime();
+
+        // Use appropriate signature algorithm based on your keyPair algorithm.
+        String signatureAlgorithm = "SHA256WithRSA";
+
+        SubjectPublicKeyInfo subjectPublicKeyInfo = SubjectPublicKeyInfo.getInstance(keyPair
+                .getPublic().getEncoded());
+
+        X509v3CertificateBuilder certificateBuilder = new X509v3CertificateBuilder(dnName,
+                certSerialNumber, startDate, endDate, dnName, subjectPublicKeyInfo);
+
+        List<GeneralName> altNames = new ArrayList<GeneralName>();
+        for (String name : genericNameList) {
+            if (name.matches(DNS_PATTERN)) {
+                Log.info(c, "selfSign", "adding dns GeneralName " + name);
+                altNames.add(new GeneralName(GeneralName.dNSName, name));
+            } else if (name.matches(IPV4_PATTERN)) {
+                Log.info(c, "selfSign", "adding ip GeneralName " + name);
+                altNames.add(new GeneralName(GeneralName.dNSName, name));
+            }
+        }
+
+        GeneralNames subjectAltNames = GeneralNames.getInstance(new DERSequence((GeneralName[]) altNames.toArray(new GeneralName[] {})));
+        certificateBuilder.addExtension(Extension.subjectAlternativeName, false, subjectAltNames);
+
+        ContentSigner contentSigner = new JcaContentSignerBuilder(signatureAlgorithm).setProvider(
+                bcProvider).build(keyPair.getPrivate());
+
+        X509CertificateHolder certificateHolder = certificateBuilder.build(contentSigner);
+
+        Certificate selfSignedCert = new JcaX509CertificateConverter()
+                .getCertificate(certificateHolder);
+
+        return selfSignedCert;
+    }
+
+    public static void exportPublicKeyToFile(String path, KeyPair keyPair) throws Exception {
+        Log.info(c, "exportPublicKeyToFile", "entering SSLUtils.exportPublicKeyToFile");
+
+        PublicKey publicKey = keyPair.getPublic();
+        File file = new File(path);
+        file.getParentFile().mkdirs();
+        file.createNewFile();
+
+        FileWriter fw = new FileWriter(file, false);
+        JcaPEMWriter pemWriter = new JcaPEMWriter(fw);
+        pemWriter.writeObject(publicKey);
+        Log.info(c, "exportPublicKeyToFile", "exported to  " + path);
+    }
+
+    public static void exportPrivateKeyToFile(String path, KeyPair keyPair) throws Exception {
+        Log.info(c, "exportPrivateKeyToFile", "entering SSLUtils.exportPrivateKeyToFile");
+
+        PrivateKey privateKey = keyPair.getPrivate();
+        File file = new File(path);
+        file.getParentFile().mkdirs();
+        file.createNewFile();
+
+        FileWriter fw = new FileWriter(file, false);
+        JcaPEMWriter pemWriter = new JcaPEMWriter(fw);
+        pemWriter.writeObject(privateKey);
+
+        pemWriter.close();
```

Comment:
As above, use a try-with-resources block to ensure that the file stream is always closed.